FieldSolver.cu

```diff
@@ -0,0 +1,287 @@
+#include "FieldSolver.cuh"
+
+
+
+
+__constant__ double3 d_cell_size[1];
+__constant__ int3 d_n_nodes[1];
+
+__constant__ double dev_dxdxdydy[1];
+__constant__ double dev_dxdxdzdz[1];
+__constant__ double dev_dydydzdz[1];
+__constant__ double dev_dxdxdydydzdz[1];
+
+__constant__ int dev_end[1];
+
+__device__ int GetIdx() {
+	//int xStepthread = 1;
+	int xStepBlock = blockDim.x;
+	int yStepThread = d_n_nodes[0].x;
+	int yStepBlock = yStepThread * blockDim.y;
+	int zStepThread = d_n_nodes[0].x * d_n_nodes[0].y;
+	int zStepBlock = zStepThread * blockDim.z;
+	return (threadIdx.x + blockIdx.x * xStepBlock) +
+		(threadIdx.y * yStepThread + blockIdx.y * yStepBlock) +
+		(threadIdx.z * zStepThread + blockIdx.z * zStepBlock);
```

Comment:
Насколько я понимаю, эта функция ставит в соответствие нити на видеокарте узел в spat_mesh? Что произойдет, если узлов в spat_mesh больше, чем общее количество нитей, запрошенное при вызове ядра?

Reply:
Эта функция предназначена для операций по всему объему, может быть скорректированно для запуска только для неграничных узлов. Сейчас количество узлов вдоль одной из осей должно быть кратно 16 (см. GetThreads() GetBlocks(threads)). Для случая описанного мной выше (запуск ядра только для неграничных узлов размер стороны меша будет (n*16 +2)
Число 16 может быть выбрано иным для оптимизации размера блока на конкретной видеокарте. так же с меньшим размером блоков потом проще оптимизировать с использованием __shared__ памяти устройства

Reply:
Хорошо, тогда такой вопрос:
CPU-версия могла работать с сетками с любым количеством узлов (вернее, сколько мог выдать malloc).  Например, 
```ini
[Spatial mesh]
grid_x_size = 5.0
grid_x_step = 0.1
grid_y_size = 5.0
grid_y_step = 0.1
grid_z_size = 10.0
grid_z_step = 0.1
```
такая секция в конфиге задает сетку 50x50x100 (ну или что-то около). 


Какие ограничения на количество узлов, которые пользователь может указать в конфиге, есть в GPU-версии?
Насколько я понимаю,
1) число внутренних узлов по каждой оси должно быть кратно 16
2) общее число узлов должно быть меньше максимально возможного числа одновременно запущенных процессов на видеокарте, которое определяется свойствами
```
(14) Multiprocessors, ( 32) CUDA Cores/MP:     448 CUDA Cores
Maximum number of threads per multiprocessor:  1536
```
см. , например, 
https://devtalk.nvidia.com/default/topic/978550/cuda-programming-and-performance/maximum-number-of-threads-on-thread-block/
или
https://stackoverflow.com/questions/31735138/cuda-threads-per-block-limitation


Нужно проверять только эти 2 ограничения или есть еще какие-то?

Reply:
Основное ограничение - память устройства, что в дальнейшем может привести к избавлению от хранения на постоянной основе полей (вполне верояно что в GPU коде дешевле будет считать).


"1. число внутренних узлов по каждой оси должно быть кратно 16" 
в данном случае :
1. общее число узлов по каждой оси должно быть кратно 16 
( это будет относиться к внутренним узлам в случае корректировки)

это ограничение происходит от реализации. из-за параметра:
 `Warp size:                                     32`
cuda runtime плохо справляется с разбиением большой задачи на warp'ы которые единовременно исполняет мультипроцессор. и на этом теряется много времени.

даже если от этого избавиться, ограничение "кратности размера сетки числу n" весьма вероятно возникнет в том или ином виде из-за ограничения технического:
"Maximum number of threads per block"
так как одним блоком большие сетки не рассчитать и разбивать на блоки поменьше все равно придется.

так же не на прямую это связанно с параметрами
`  Maximum number of threads per block:           1024 
  Max dimension size of a thread block (x,y,z): (1024, 1024, 64)
  Max dimension size of a grid size    (x,y,z): (65535, 65535, 65535)`
обсуждаемыми по данным ссылкам. 
последние два умножив скалярно  (1024, 1024, 64)*(65535, 65535, 65535) - получим предел общего числа узлов которое можно обсчитать одним вызовом __global__ метода. 
Однако память устройства закончится куда раньше.

`Maximum number of threads per multiprocessor:  1536`
честно говоря требует небольшого исследования. О чем этот параметр на самом деле говорит, так как запускал __global__ ядра на сетку общим размером 450*450*450 еще в Cuda версии 6.  (то есть если задать количество нитей очевидно больше 1536 в разы)

Reply:
> последние два умножив скалярно (1024, 1024, 64)*(65535, 65535, 65535) - получим предел общего числа узлов которое можно обсчитать одним вызовом global метода.

Похоже, все не так оптимистично. Есть максимальное число процессов на блок и есть ограничение на величину индекса по оси блока. При этом произведение всех индексов в блоке должно быть меньше указанного максимального числа процессов на блок (см. цитату из одной из ссылок выше). 

Остается вопрос, сколько блоков можно запустить на устройстве. 

> 1. The maximum number of threads in the block is limited to 1024. This is the product of whatever your threadblock dimensions are (x*y*z). For example (32,32,1) creates a block of 1024 threads. (33,32,1) is not legal, since 33*32*1 > 1024.
> 2. The maximum x-dimension is 1024. (1024,1,1) is legal. (1025,1,1) is not legal.
> 3. The maximum y-dimension is 1024. (1,1024,1) is legal. (1,1025,1) is not legal.
> 4. The maximum z-dimension is 64. (1,1,64) is legal. (2,2,64) is also legal. (1,1,65) is not legal.
> Your choice of threadblock dimensions (x,y,z) must satisfy *each* of the rules 1-4 above.


Ладно, все эти вопросы с количеством узлов пока можно оставить как есть. Мне надо будет только не забыть потом об этом.